oak/server/oak_node.cc

```diff
@@ -143,4 +147,29 @@ bool OakNode::WaitOnChannels(std::vector<std::unique_ptr<ChannelStatus>>* status
   }
 }
 
+void OakNode::AddLabelsToMessage(Message* message) const {
+  absl::ReaderMutexLock lock(&labels_mu_);
+  message->labels.secrecy_tags.insert(labels_.secrecy_tags.begin(), labels_.secrecy_tags.end());
+  message->labels.integrity_tags.insert(labels_.integrity_tags.begin(),
+                                        labels_.integrity_tags.end());
+}
+
+void OakNode::UpdateLabelsFromMessage(const Message* message) {
+  absl::MutexLock lock(&labels_mu_);
+  // Any secrecy tags from the message are added to the Node's current secrecy
+  // label.
+  labels_.secrecy_tags.insert(message->labels.secrecy_tags.begin(),
+                              message->labels.secrecy_tags.end());
+
+  // Any of the Node's current integrity tags which are absent from the message
+  // are removed from the Node's current integrity label.
```

Comment:
This does not sound right, but I need to think more about it...

Reply:
I think the confusion is caused by the "floating" label mechanism, since it updates the labels based on the input message. The more expected semantics would be that labels attached to nodes don't implicitly change when reading messages, rather nodes have to explicitly request changing their labels. So integrity is a requirement for a node on its inputs, while secrecy is on its outputs. In the corresponding way, sending a message to, say, the log node, should not be allowed if the message has a non-empty secrecy label.

Also  note the following, from the paper:
> a process p with v ∈ Ip cannot read from files or processes that lack v integrity

Reply:
Yeah, this PR is doing label propagation/tracking rather than policing, so it's almost certainly not going be what we end up with &ndash; in which case it may not be worth putting in something that's going to be radically redesigned.